pkg/controller/virtualmachineimport/virtualmachineimport_controller.go

```diff
@@ -117,25 +128,63 @@ func (r *ReconcileVirtualMachineImport) Reconcile(request reconcile.Request) (re
 	}
 
 	// Connect to oVirt:
-	ovirtSecret, err := r.fetchOvirtSecret(instance)
+	ovirtSecretObj, err := r.fetchOvirtSecret(instance)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
-	ovirtConnection, err := createOvirtConnection(ovirtSecret)
+
+	ovirtSecretDataMap := make(map[string]string)
+	err = yaml.Unmarshal(ovirtSecretObj.Data[ovirtSecretKey], &ovirtSecretDataMap)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
 
+	if _, ok := ovirtSecretDataMap["caCert"]; !ok {
+		return reconcile.Result{}, fmt.Errorf("oVirt secret must contain caCert attribute")
+	}
+	if len(ovirtSecretDataMap["caCert"]) == 0 {
+		return reconcile.Result{}, fmt.Errorf("oVirt secret caCert cannot be empty")
+	}
+	caCert, err := base64.StdEncoding.DecodeString(ovirtSecretDataMap["caCert"])
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("Failed to decode oVirt secret caCert to base64 format")
+	}
+	ovirt, err := ovirtclient.NewRichOvirtClient(
+		&ovirtclient.ConnectionSettings{
+			URL:      ovirtSecretDataMap["apiUrl"],
+			Username: ovirtSecretDataMap["username"],
+			Password: ovirtSecretDataMap["password"],
+			CACert:   caCert,
+		},
+	)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	defer ovirt.Close()
+
 	// Fetch oVirt VM:
-	vm, err := fetchOvirtVM(ovirtConnection, &instance.Spec.Source.Ovirt.VM)
+	sourceVMID := instance.Spec.Source.Ovirt.VM.ID
+	sourceVMName := instance.Spec.Source.Ovirt.VM.Name
+	var sourceVMClusterName *string
+	var sourceVMClusterID *string
+	if instance.Spec.Source.Ovirt.VM.Cluster != nil {
+		sourceVMClusterName = instance.Spec.Source.Ovirt.VM.Cluster.Name
+		sourceVMClusterID = instance.Spec.Source.Ovirt.VM.Cluster.ID
+	}
+	vm, err := ovirt.GetVM(sourceVMID, sourceVMName, sourceVMClusterName, sourceVMClusterID)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
 
-	diskAttachmentsLink, _ := vm.DiskAttachments()
-	diskAttachments, _ := ovirtConnection.FollowLink(diskAttachmentsLink)
-	dvs := createDVmap(diskAttachments.(*ovirtsdk.DiskAttachmentSlice), ovirtSecret)
-	vmSpec := newVMForCR(vm, dvs, diskAttachments.(*ovirtsdk.DiskAttachmentSlice))
+	// Stop VM
+	vmID, _ := vm.Id()
+	err = ovirt.StopVM(vmID)
```

Comment:
We need to memorize original vm status maybe as annotation. We need it to restore status if import operation fails.

Reply:
In the discussion over the document we concluded the vm status shouldn't be changed on failure at provider's end.
Meaning we shouldn't start it if the process failed. I'll verify this point.